From f3581a48826c3cc0a1d4927ee1dc5c44ff5e61bf Mon Sep 17 00:00:00 2001
From: Dmitry Ivanov <idmitry@qti.qualcomm.com>
Date: Thu, 15 Apr 2021 15:35:26 +0200
Subject: [PATCH] formats: Handle potential GL error code after
 glGetIntegerv(GL_IMPLEMENTATION_COLOR_READ_TYPE).

OpenGL/ES doesn't allow reading from DEPTH/STENCIL type buffers, only
from the color attachments. This also means the glReadBuffer() default
setting goes to COLOR_ATTACHMENT0, which conflicts if that is being
used together with a Framebuffer/FBO which has only a DEPTH/STENCIL
type attachment, i.e. there is no image assigned to any of the color
attachments, let alone the first one. According to the specification
this will trigger a GL error being generated when calling
glGetIntegerv() with GL_IMPLEMENTATION_COLOR_READ_TYPE:
GL_INVALID_OPERATION (= 0x0502).

This type of FBO is being tested for readback checks in virglrenderer,
but no GL error check is done. It functions similar to errno, in that
an error code sticks around up until the next time it is being read,
which in this case is an assert() call against the GL error code being
GL_NO_ERROR. That fails, causing an assert() induced exit of the
entire coqos-virtio-gpu-virgl application.

NOTE: The issue has been reported upstream, and, if acknowledged as an
      issue of virglrenderer, will eventually get fixed upstream.

Upstream ticket: https://gitlab.freedesktop.org/virgl/virglrenderer/-/issues/213

Signed-off-by: Dmitry Ivanov <idmitry@qti.qualcomm.com>
---
 src/vrend_formats.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/vrend_formats.c b/src/vrend_formats.c
index 670b3a40..63a0cb59 100644
--- a/src/vrend_formats.c
+++ b/src/vrend_formats.c
@@ -367,6 +367,7 @@ static struct vrend_format_table gles_bgra_formats[] = {
 static bool color_format_can_readback(struct vrend_format_table *virgl_format, int gles_ver)
 {
    GLint imp = 0;
+   GLenum status = GL_NO_ERROR;
 
    if (virgl_format->format == VIRGL_FORMAT_R8G8B8A8_UNORM)
       return true;
@@ -391,7 +392,10 @@ static bool color_format_can_readback(struct vrend_format_table *virgl_format, i
 
    /* Check implementation specific readback formats */
    glGetIntegerv(GL_IMPLEMENTATION_COLOR_READ_TYPE, &imp);
-   if (imp == (GLint)virgl_format->gltype) {
+   /* GLES can throw errors for certain combinations which are ok for normal GL,
+      so catch them here */
+   status = glGetError();
+   if (status == GL_NO_ERROR && imp == (GLint)virgl_format->gltype) {
       glGetIntegerv(GL_IMPLEMENTATION_COLOR_READ_FORMAT, &imp);
       if (imp == (GLint)virgl_format->glformat)
          return true;
-- 
2.31.1

